sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java

```diff
@@ -117,21 +118,27 @@ public long id() {
         URI address = addressUri.getURI();
 
         final RntbdRequestArgs requestArgs = new RntbdRequestArgs(request, address);
+
         requestArgs.traceOperation(logger, null, "invokeStoreAsync");
 
         final RntbdEndpoint endpoint = this.endpointProvider.get(address);
         final RntbdRequestRecord record = endpoint.request(requestArgs);
 
         logger.debug("RntbdTransportClient.invokeStoreAsync({}, {}): {}", address, request, record);
 
-        return Mono.fromFuture(record).doFinally(signalType -> {
-            logger.debug("SignalType.{} received from reactor: {\n  endpoint: {},\n  record: {}\n}",
-                signalType.name(),
-                endpoint,
-                record);
-            if (signalType == SignalType.CANCEL) {
-                record.stage(RntbdRequestRecord.Stage.CANCELLED_BY_CLIENT);
+        return Mono.fromFuture(record.whenComplete((response, throwable) -> {
+
+            record.stage(RntbdRequestRecord.Stage.COMPLETED);
+
+            if (throwable == null) {
+                response.setRequestTimeline(record.takeTimelineSnapshot());
+            } else if (throwable instanceof CosmosClientException) {
```

Comment:
Reactor might / will wrap the exception in ReactiveException. 
The safest way to check `throwable instanceOf CosmosClientException` is this. 
```
final Throwable unwrappedThrowable = reactor.core.Exceptions.unwrap(throwable);
if (unwrappedThrowable instanceof CosmosClientException) {
 .....
}
```

Reply:
@kushagraThapar `reactor.core.Exceptions.unwrap` should not be needed in this layer. right?

Reply:
Correct: The `CosmosClientException` originates in the Direct TCP stack and so `reactor.core.Exceptions.unwrap` is not needed.